platform/src/main/java/org/hillview/dataset/remoting/HillviewServer.java

```diff
@@ -102,26 +106,62 @@ public HillviewServer(final HostAndPort listenAddress, final IDataSet dataSet) t
                                         .maxMessageSize(MAX_MESSAGE_SIZE)
                                         .build()
                                         .start();
-        this.dataSets = new ConcurrentHashMap<Integer, IDataSet>();
-        this.put(this.dsIndex.getAndIncrement(), dataSet);
+        this.dataSets = CacheBuilder.<Integer, IDataSet>newBuilder()
+                .expireAfterAccess(expireTimeInHours, TimeUnit.HOURS)
+                .removalListener(
+                        (RemovalListener<Integer, IDataSet>) removalNotification ->
+                                HillviewLogger.instance.info("Removing reference to dataset", "{0}: {1}",
+                                    removalNotification.getKey(), removalNotification.getValue().toString()))
+                .build();
     }
 
-    synchronized private IDataSet get(int index) {
-        return this.dataSets.get(index);
-    }
+    //////////////////////
 
-    synchronized private void put(int index, IDataSet dataSet) {
+    synchronized private int save(IDataSet dataSet) {
+        int index = this.dsIndex.getAndIncrement();
         HillviewLogger.instance.info("Inserting dataset", "{0}", index);
         this.dataSets.put(index, dataSet);
+        return index;
     }
 
-    synchronized private void remove(int index) {
-        HillviewLogger.instance.info("Removing dataset", "{0}", index);
-        this.dataSets.remove(index);
+    /**
+     * Retrieve the dataset with the specified index.
+     * @param index     Dataset index.
+     * @param observer  Observer that is notified if the dataset is not available.
+     * @return          The dataset, or null if there is no such dataset.
+     */
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
```

Comment:
Looks like this is not required because there is no use of inverted booleans in the method